### DISCUSSION TITLE
spindles dection issue

### DISCUSSION THREAD

**Discussion by chensnbetter** [raphaelvallat/yasa#102] [Q&A] [answered]
When trying to detect spindles, it always shows that no spindles were found in data.
<img width="427" alt="error" src="https://user-images.githubusercontent.com/115486173/202891898-da562d84-2bc7-4ea9-97dd-5e6489ab8995.png">

import mne
raw = mne.io.read_raw_edf('C://Users//ASUS//Desktop//yasa_example_night_young.edf', preload=True)
raw
print(raw.ch_names)
raw.drop_channels(['ROC-A1', 'LOC-A2', 'EMG1-EMG2', 'EKG-R-EKG-L'])
chan = raw.ch_names
print(chan)
print(raw.info['sfreq'])
raw.resample(100)
sf = raw.info['sfreq']
sf
raw.filter(0.3, 45)
data = raw.get_data(units="uV")
print(data.shape)
import pandas as pd
hypno = pd.read_csv("C://Users//ASUS//Desktop//yasa_example_night_young_hypno.csv", squeeze=True)
hypno
import yasa
yasa.plot_hypnogram(hypno);
yasa.sleep_statistics(hypno, sf_hyp=1/30)
counts, probs = yasa.transition_matrix(hypno)
probs.round(3)
import numpy as np
np.diag(probs.loc[2:, 2:]).mean().round(3)
hypno_up = yasa.hypno_upsample_to_data(hypno, sf_hypno=1/30, data=raw)
print(len(hypno_up))
yasa.plot_spectrogram(data[chan.index("C4-A1")], sf, hypno_up);
yasa.bandpower(raw)
yasa.bandpower(raw, relative=False, bands=[(1, 9, "Slow"), (9, 30, "Fast")])
bandpower = yasa.bandpower(raw, hypno=hypno_up, include=(2, 3, 4))
bandpower.to_csv("bandpower.csv")
sp = yasa.spindles_detect(raw, hypno=hypno_up, include=(2, 3))

It didn't go well while doing it with data other than the sample data of yasa. So I tried it with the sample data of yasa, and the same error appears.

**Comment by remrama** [comment]
Hi @chensnbetter . I just tried replicating this issue but everything seemed to work fine for me. The first thing I did was trim the code down to the essentials needed for the spindle detection and ran this: 
```python
import mne
import pandas as pd
import yasa

raw_fname = "./yasa_example_night_young.edf"
hypno_fname = "./yasa_example_night_young_hypno.csv"
raw = mne.io.read_raw_edf(raw_fname, preload=True)
raw.drop_channels(['ROC-A1', 'LOC-A2', 'EMG1-EMG2', 'EKG-R-EKG-L'])
raw.resample(100)
raw.filter(0.3, 45)
hypno = pd.read_csv(hypno_fname, squeeze=True)
hypno_up = yasa.hypno_upsample_to_data(hypno, sf_hypno=1/30, data=raw)
sp = yasa.spindles_detect(raw, hypno=hypno_up, include=(2, 3))
```

That worked well for me, spindles were detected. 

Then I thought maybe some of the other functions might have accidentally been modifying `raw` or something (mne can be sneaky with the in-place modifications). So I reran the script including all your other lines (except for the prints) and still, spindles were detected.

Could you try re-running this with the trimmed down code? Make sure you're up to date on YASA (I'm using YASA `v0.6.2` in a Python `v3.10.6` environment), and also be sure to exit and re-open a new iPython session if you're using something like that (to clear out any existing variables). Probably download fresh copies of the sample data to be sure as well.

**Reply by chensnbetter** [reply]
Thanks @remrama. I reran this with the trimmed down code. The same error appeared when I ensured YASA v0.6.2 in a Python v3.10.6 environment. 

In addition, I want to know whether it has something to do with MNE's version (I'm using MNE 1.2.2)?

**Reply by raphaelvallat** [reply]
Hi @chensnbetter. Could you please retry after downgrading to MNE 1.1, MNE 1.0 and MNE 0.24? Thanks

**Reply by chensnbetter** [reply]
It didn't work to change the version.

**Reply by raphaelvallat** [reply]
Can you please run `pip list` and copy-paste the results here? Thanks

**Reply by chensnbetter** [reply]
Package            Version
------------------ ---------
antropy            0.1.4
appdirs            1.4.4
asttokens          2.1.0
backcall           0.2.0
certifi            2022.9.24
charset-normalizer 2.1.1
colorama           0.4.6
contourpy          1.0.6
cycler             0.11.0
debugpy            1.6.3
decorator          5.1.1
entrypoints        0.4
executing          1.2.0
fonttools          4.38.0
fsspec             2022.11.0
idna               3.4
ipykernel          6.17.1
ipython            8.6.0
ipywidgets         8.0.2
jedi               0.18.2
Jinja2             3.1.2
joblib             1.2.0
jupyter_client     7.4.7
jupyter_core       5.0.0
jupyterlab-widgets 3.0.3
kiwisolver         1.4.4
littleutils        0.2.2
llvmlite           0.39.1
lspopt             1.2.0
MarkupSafe         2.1.1
matplotlib         3.6.2
matplotlib-inline  0.1.6
mne                1.2.2
nest-asyncio       1.5.6
numba              0.56.4
numpy              1.23.5
outdated           0.2.2
packaging          21.3
pandas             1.5.2
parso              0.8.3
pickleshare        0.7.5
Pillow             9.3.0
pip                22.3.1
platformdirs       2.5.4
pooch              1.6.0
prompt-toolkit     3.0.33
psutil             5.9.4
pure-eval          0.2.2
Pygments           2.13.0
pyparsing          3.0.9
pyriemann          0.3
python-dateutil    2.8.2
pytz               2022.6
pywin32            305
PyYAML             6.0
pyzmq              24.0.1
requests           2.28.1
scikit-learn       1.1.3
scipy              1.9.3
seaborn            0.12.1
setuptools         63.2.0
six                1.16.0
sleepecg           0.5.2
stack-data         0.6.1
stochastic         0.7.0
tensorpac          0.6.5
threadpoolctl      3.1.0
tornado            6.2
tqdm               4.64.1
traitlets          5.5.0
urllib3            1.26.13
wcwidth            0.2.5
widgetsnbextension 4.0.3
yasa               0.6.2

I have reinstalled python and other related installation packages. And that is also useless.

**Reply by raphaelvallat** [reply]
Thanks for this. Can you please share here your Jupyter notebook? Please make sure to add a cell at the end with:

```python
print(raw)
print(hypno)
sp.summary()
```

**Reply by chensnbetter** [reply]
<img width="553" alt="屏幕截图 2022-11-27 225724" src="https://user-images.githubusercontent.com/115486173/204141974-2b28e6f3-e706-47e9-8fe3-ad9e7f9c7eb9.png">
<img width="889" alt="屏幕截图 2022-11-27 225655" src="https://user-images.githubusercontent.com/115486173/204141977-1af35b26-4f0a-4b20-97e2-60af05d7a3cb.png">
<img width="708" alt="屏幕截图 2022-11-27 225806" src="https://user-images.githubusercontent.com/115486173/204141979-c4cb5704-433f-4372-a2d5-fcdbe0d7f453.png">

**Reply by raphaelvallat** [reply]
I really have no idea what's going on unfortunately 😕 Can you try running the other YASA example notebooks here? https://github.com/raphaelvallat/yasa/tree/master/notebooks

**Reply by chensnbetter** [reply]
Thanks, I will try the other notebooks.
Meanwhile, I would like to ask, is there any problem in the following process?
<img width="449" alt="屏幕截图 2022-11-28 204135" src="https://user-images.githubusercontent.com/115486173/204579842-7b2ae194-0d9d-4f0e-821a-e287f1ca3645.png">

**Comment by chensnbetter** [comment]
Strangely, why can I detect slow waves successfully?

**Reply by raphaelvallat** [reply]
Hmmm, there must be something specific about the steps/sub-functions in the spindles_detect function

**Comment by raphaelvallat** [comment, marked as answer]
Opened an issue here: https://github.com/raphaelvallat/yasa/issues/107 Let's move the conversation there
